api/src/main/java/org/openmrs/validator/ObsValidator.java

```diff
@@ -252,5 +258,129 @@ else if (!isObsGroup) {
+	 *     <li>Validates if Obs' numeric value is within the valid range; i.e. >= low absolute && <= high absolute.</li>
+	 *     <li>Sets field errors if numeric value is outside the valid range</li>
+	 * <ol/>
+	 *
+	 * @param obs Observation to validate
+	 * @param errors Errors to record validation issues
+	 *               
+	 * @since 2.7.0
+	 */
+	private void validateConceptReferenceRange(Obs obs, Errors errors) {
+		ConceptReferenceRange conceptReferenceRange = evaluateReferenceRange(obs.getConcept(), obs.getPerson());
+
+		if (conceptReferenceRange != null) {
+			validateAbsoluteRanges(obs, conceptReferenceRange, errors);
+			setObsReferenceRange(obs, conceptReferenceRange);
+		}
+	}
+
+	/**
+	 * Evaluates the criteria and return the most strict {@link ConceptReferenceRange} for a given concept.
+	 * It considers all valid ranges that match the criteria for the person.
+	 *
+	 * @param concept The concept to evaluate
+	 * @param person The person for whom the range is being evaluated
+	 * @return The strictest {@link ConceptReferenceRange}, or null if no valid range is found
+	 * 
+	 * @since 2.7.0
+	 */
+	public static ConceptReferenceRange evaluateReferenceRange(Concept concept, Person person) {
+		if (concept != null && concept.getDatatype() != null && concept.getDatatype().isNumeric()) {
+			List<ConceptReferenceRange> referenceRanges = Context.getConceptService()
+				.getConceptReferenceRangesByConceptId(concept.getConceptId());
+
+			if (!referenceRanges.isEmpty()) {
+				ConceptReferenceRangeUtility utility = new ConceptReferenceRangeUtility();
+				List<ConceptReferenceRange> validRanges = new ArrayList<>();
+
+				for (ConceptReferenceRange referenceRange : referenceRanges) {
+					if (utility.evaluateCriteria(referenceRange.getCriteria(), person)) {
+						validRanges.add(referenceRange);
+					}
+				}
+
+				if (!validRanges.isEmpty()) {
+					return findStrictestReferenceRange(validRanges);
+				}
+			}
+		}
+		return null;
+	}
+	
+	/**
+	 * Finds the strictest {@link ConceptReferenceRange} from a list of valid ranges.
+	 * The strictest range is determined by having the highest lower bound and the lowest upper bound.
+	 *
+	 * @param conceptReferenceRanges A list of valid {@link ConceptReferenceRange} objects
+	 * @return The strictest {@link ConceptReferenceRange} constructed from the strictest bounds
+	 * 
+	 * @since 2.7.0
+	 */
+	public static ConceptReferenceRange findStrictestReferenceRange(List<ConceptReferenceRange> conceptReferenceRanges) {
+		Double strictestLowAbsolute = conceptReferenceRanges.stream()
+			.map(ConceptReferenceRange::getLowAbsolute)
+			.filter(Objects::nonNull)
+			.max(Double::compareTo)
+			.orElse(null);
+
+		Double strictestHiAbsolute = conceptReferenceRanges.stream()
+			.map(ConceptReferenceRange::getHiAbsolute)
+			.filter(Objects::nonNull)
+			.min(Double::compareTo)
+			.orElse(null);
+
+		ConceptReferenceRange strictestRange = new ConceptReferenceRange();
```

Comment:
Why are we creating a new concept reference range for return? Shouldn't we pick from the given conceptReferenceRanges list parameter?

Reply:
The reason we are creating a "new" `ConceptReferenceRange` is because the ranges could be from different rerenceRanges. 
e.g.
`ConceptReferenceRange` one has a range of `80-150`.
`ConceptReferenceRange` two has a range of `60 - 140`.

The "strictest" range will be `80-140`. This range is deduced from the 2 reference ranges. Hence, we shouldn't return referenceRange from the list of `conceptReferenceRanges`.

Reply:
Makes perfect sense! 👍

Reply:
Can you add a comment in the code?